docs/design.rst

```diff
@@ -0,0 +1,403 @@
+stores and returns a :class:`~asyncio.Future` which will be completed when a
+matching pong frame is received.
+
+:meth:`~protocol.WebSocketCommonProtocol.pong()` writes a pong frame.
+
+:meth:`~protocol.WebSocketCommonProtocol.close()` writes a close frame and
+waits for the connection to terminate.
+
+Outgoing data is written to a :class:`~asyncio.StreamWriter` in order to
+implement flow control and provide backpressure from the TCP connection.
+
+.. _closing-handshake:
+
+Closing handshake
+.................
+
+When the other side of the connection initiates the closing handshake,
+:meth:`~protocol.WebSocketCommonProtocol.read_message()` receives a close
+frame while in the ``OPEN`` state. It moves to the ``CLOSING`` state, sends a
+close frame, and returns ``None``, causing
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate.
+
+When this side of the connection initiates the closing handshake with
+:meth:`~protocol.WebSocketCommonProtocol.close()`, it moves to the ``CLOSING``
+state and sends a close frame. When the other side sends a close frame,
+:meth:`~protocol.WebSocketCommonProtocol.read_message()` receives it in the
+``CLOSING`` state and returns ``None``, also causing
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate.
+
+If the other side doesn't send a close frame within the connection's timeout,
+:meth:`~protocol.WebSocketCommonProtocol.close()` cancels
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task`, which has the
+same effect.
+
+Then ``websockets`` terminates the TCP connection.
+
+
+.. _connection-termination:
+
+Connection termination
+----------------------
+
+:attr:`~protocol.WebSocketCommonProtocol.close_connection_task`, which is
+started when the WebSocket connection is established, is responsible for
+eventually closing the TCP connection.
+
+First :attr:`~protocol.WebSocketCommonProtocol.close_connection_task` waits
+for :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` to terminate,
+which may happen as a result of:
+
+- a successful closing handshake: as explained above, this exits the infinite
+  loop in :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task`;
+- a timeout while waiting for the closing handshake to complete: this cancels
+  :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task`;
+- a protocol error, including connection errors: depending on the exception,
+  :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` fails the
+  WebSocket connection with a suitable code and exits.
+
+:attr:`~protocol.WebSocketCommonProtocol.close_connection_task` is separate
+from :attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` because that
+makes it easier to implement the timeout on the closing handshake. Cancelling
+:attr:`~protocol.WebSocketCommonProtocol.transfer_data_task` creates no risk
+of cancelling :attr:`~protocol.WebSocketCommonProtocol.close_connection_task`
+and thus failing to close the TCP connection.
```

Comment:
For these two points, why can't you just pass the timeout when you start closing the handshake? And for cancelling the data transfer task, you can still shield or handle `CancelledError` for closing to happen correctly. IIRC, this seems similar to the structure you had before, but I don't see why you couldn't incorporate the logic you needed.

Reply:
It was very hard to convince oneself that the previous implementation was robust in the face of an arbitrary sequence of the following events:

- calling `close()`, possibly multiple times (that shouldn't raise an error)
- receiving a close frame from the other end, which initiates the closing handshake
- receiving a TCP half-close or a TCP close from the other end

#112 is perhaps the best example: it's very hard to figure out what happens in such cases.

The new design has much simpler invariants: there's a well defined and easily auditable contract for the two tasks:
- when they run
- when they can be cancelled
- whether they can exit with an exception (no)

Since the two tasks essentially run in sequence — `close_connection_task` waits for `transfer_data_task` to terminate before doing anything — it must be possible to:
- concatenate them
- complete a future at the junction
- await that future instead of awaiting the first task
- checking the value of that future whenever cancellation may propagate the task and let it propagate only if the futur isn't completed yet

I think that would make the code more complex and that's why I stuck with two tasks.

(I'm wondering how much memory overhead this creates. I'm planning to check that. But it won't be much compared to buffers and compression state.)